File: android/app/src/main/java/com/smartcardpoc/FutronicSdkWrapper.java
package com.smartcardpoc;

import javax.smartcardio.*;
import java.util.regex.Pattern;

public class FutronicSdkWrapper {
    private CardTerminal terminal;
    private Card card;
    private CardChannel channel;

    public boolean open(CardTerminal cardTerminal) {
        try {
            this.terminal = cardTerminal;
            return true;
        } catch (Exception e) {
            e.printStackTrace();
            return false;
        }
    }

    public boolean isReaderOpen() {
        return terminal != null;
    }

    public byte[] transmit(String command) {
        try {
            if (!terminal.isCardPresent()) {
                throw new CardException("No card present in the reader. Please insert a CPR card.");
            }

            if (card == null) {
                try {
                    card = terminal.connect("T=1");
                } catch (CardException e1) {
                    try {
                        card = terminal.connect("T=0");
                    } catch (CardException e2) {
                        throw new CardException("Failed to connect with both T=1 and T=0 protocols.", e2);
                    }
                }
            }

            CardChannel currentChannel = card.getBasicChannel();
            if (currentChannel == null) {
                throw new CardException("Could not get a valid card channel.");
            }

            byte[] commandBytes = stringToBytes(command);
            CommandAPDU commandAPDU = new CommandAPDU(commandBytes);
            ResponseAPDU responseAPDU = currentChannel.transmit(commandAPDU);

            return responseAPDU.getBytes();
        } catch (Exception e) {
            close();
            if (e instanceof RuntimeException) {
                throw (RuntimeException) e;
            } else {
                throw new RuntimeException(e);
            }
        }
    }

    public void close() {
        try {
            if (card != null) {
                card.disconnect(false);
            }
        } catch (Exception e) {
            // Ignore
        }
        terminal = null;
        card = null;
        channel = null;
    }

    private byte[] stringToBytes(String s) {
        String cleanString = s.replaceAll("[^0-9a-fA-F]", "");
        if (cleanString.length() % 2 != 0) {
            throw new IllegalArgumentException("Hex string must have an even number of characters");
        }
        byte[] bytes = new byte[cleanString.length() / 2];
        for (int i = 0; i < cleanString.length(); i += 2) {
            bytes[i / 2] = (byte) Integer.parseInt(cleanString.substring(i, i + 2), 16);
        }
        return bytes;
    }
}

File: android/app/src/main/java/com/smartcardpoc/FutronicSdkWrapper.kt
package com.smartcardpoc

import javax.smartcardio.Card
import javax.smartcardio.CardChannel
import javax.smartcardio.CardException
import javax.smartcardio.CardTerminal
import javax.smartcardio.CommandAPDU
import javax.smartcardio.ResponseAPDU
import android.util.Log

class FutronicSdkWrapper {
    private var terminal: CardTerminal? = null
    private var card: Card? = null
    private var channel: CardChannel? = null

    fun open(cardTerminal: CardTerminal): Boolean {
        Log.d("FutronicSdkWrapper", "Opening terminal")
        return try {
            this.terminal = cardTerminal
            true
        } catch (e: Exception) {
            Log.e("FutronicSdkWrapper", "Open error: ${e.message}", e)
            false
        }
    }

    fun isReaderOpen(): Boolean = terminal != null

    fun transmit(command: String): ByteArray? {
        Log.d("FutronicSdkWrapper", "Transmitting command: $command")
        try {
            if (terminal?.isCardPresent != true) {
                throw CardException("No card present in the reader. Please insert a CPR card.")
            }

            if (card == null) {
                card?.disconnect(false)
                try {
                    card = terminal!!.connect("T=1")
                } catch (e1: CardException) {
                    try {
                        card = terminal!!.connect("T=0")
                    } catch (e2: CardException) {
                        throw CardException("Failed to connect with both T=1 and T=0 protocols.", e2)
                    }
                }
            }

            val currentChannel = card!!.basicChannel
                ?: throw CardException("Could not get a valid card channel.")

            val commandBytes = stringToBytes(command)
            val commandAPDU = CommandAPDU(commandBytes)
            val responseAPDU: ResponseAPDU = currentChannel.transmit(commandAPDU)

            return responseAPDU.bytes
        } catch (e: Exception) {
            Log.e("FutronicSdkWrapper", "Transmit error: ${e.message}", e)
            close()
            throw e
        }
    }

    fun close() {
        Log.d("FutronicSdkWrapper", "Closing wrapper")
        try {
            card?.disconnect(false)
        } catch (e: Exception) {
            Log.e("FutronicSdkWrapper", "Close error: ${e.message}", e)
        }
        terminal = null
        card = null
        channel = null
    }

    private fun stringToBytes(s: String): ByteArray {
        val cleanString = s.replace(Regex("[^0-9a-fA-F]"), "")
        if (cleanString.length % 2 != 0) {
            throw IllegalArgumentException("Hex string must have an even number of characters")
        }
        return cleanString.chunked(2).map { it.toInt(16).toByte() }.toByteArray()
    }
}